File: homework1/src/warmup.js
const rp = require('request-promise');
const crypto = require('crypto');

function change(c) {
  if (c < 0) {
    throw new RangeError('Amount Cannot be Negative');
  }
  let coins = c;
  const quarters = coins / 25;
  coins %= 25;
  const dimes = coins / 10;
  coins %= 10;
  const nickles = coins / 5;
  coins %= 5;
  return [Math.trunc(quarters), Math.trunc(dimes), Math.trunc(nickles), coins];
}

function stripQuotes(s) {
  return s.replace(/"|'/gi, '');
}

function scramble(w) {
  let words = w;
  let scrambledWord = '';
  while (words.length > 0) {
    const randomIndex = Math.trunc(Math.random() * words.length);
    scrambledWord += words.charAt(randomIndex);
    words = words.replace(words.charAt(randomIndex), '');
  }
  return scrambledWord;
}

function powers(base, limit, p) {
  let answer = 1;
  let exp = 1;
  while (answer <= limit) {
    p(answer);
    answer = base ** exp;
    exp += 1;
  }
}

function* powersGenerator(base, limit) {
  let exp = 0;
  while (base ** exp <= limit) {
    yield base ** exp;
    exp += 1;
  }
}

function say(s) {
  if (!s) {
    return '';
  }
  return function sayMore(str) {
    if (str) {
      return say(`${s} ${str}`);
    }
    return s;
  };
}

function interleave(a, ...v) {
  let counter = 0;
  const result = [];
  for (let i = 0; i < a.length; i += 1) {
    result.push(a[i]);
    if (i < v.length) {
      result.push(v[i]);
    }
    counter = i + 1;
  }
  for (let j = counter; j < v.length; j += 1) {
    result.push(v[j]);
  }
  return result;
}

function cylinder({ radius = 1, height = 1 }) {
  const surfaceArea = () => (2 * Math.PI * radius * height) + (2 * Math.PI * radius * radius);
  const volume = () => Math.PI * radius * radius * height;
  const widen = (factor) => { radius *= factor; };
  const stretch = (factor) => { height *= factor; };
  const toString = () => `Cylinder with radius ${radius} and height ${height}`;
  return Object.freeze({
    get radius() { return radius; },
    get height() { return height; },
    surfaceArea,
    volume,
    widen,
    stretch,
    toString,
  });
}

function makeCryptoFunctions(key, algorithm) {
  function encrypt(text) {
    const cipher = crypto.createCipher(algorithm, key);
    let crypted = cipher.update(text, 'utf8', 'hex');
    crypted += cipher.final('hex');
    return crypted;
  }
  function decrypt(text) {
    const decipher = crypto.createDecipher(algorithm, key);
    let dec = decipher.update(text, 'hex', 'utf8');
    dec += decipher.final('utf8');
    return dec;
  }
  return [encrypt, decrypt];
}

function randomName(params) {
  const { region, gender } = params;
  const url = 'https://uinames.com/api/';
  return rp(`${url}?amount=1&gender=${gender}&region=${region}`)
    .then(body => `${body.surname}, ${body.name}`);
}

module.exports = {
  change,
  stripQuotes,
  scramble,
  powers,
  powersGenerator,
  say,
  interleave,
  cylinder,
  makeCryptoFunctions,
  randomName,
};

File: homework2/warmup.py
from random import shuffle
import math
from Crypto.Cipher import AES
import requests


def change(c):
    if c < 0:
        raise ValueError('amount cannot be negative')
    quarters = divmod(c, 25)
    dimes = divmod(quarters[1], 10)
    nickles = divmod(dimes[1], 5)
    return quarters[0], dimes[0], nickles[0], nickles[1]


def strip_quotes(quote):
    quote = quote.replace('\'', '')
    quote = quote.replace('"', '')
    return quote


def scramble(s):
    s = list(s)
    shuffle(s)
    return ''.join(s)


def say(x=''):
    if x == '':
        return x

    def _sayMore(y=''):
        if y == '':
            return x
        return say(x + ' ' + y)
    return _sayMore


def triples(h):
    set_of_triples = []

    for a in range(3, h):
        for c in range(5, h+1):
            if c > a:
                b = math.sqrt((c ** 2) - (a ** 2))
                if b == int(b) and b > a:
                    set_of_triples.append((a, int(b), c))

    return set_of_triples


def powers(base, limit):
    exponent = 0
    current = 1
    while current <= limit:
        yield current
        exponent = exponent + 1
        current = base ** exponent


def interleave(x, *y):
    result = []
    while x or y:
        if x:
            result.append(x.pop(0))
        if y:
            result.append(y[0])
            y = y[1:]
    return result


class Cylinder:
    def __init__(self, radius=1, height=1):
        self.radius = radius
        self.height = height
        self.volume = None
        self.surface_area = None
        self.getMeasurements()

    def stretch(self, factor):
        self.height *= factor
        self.getMeasurements()
        return self

    def widen(self, factor):
        self.radius *= factor
        self.getMeasurements()
        return self

    def getMeasurements(self):
        self.volume = math.pi * (self.radius ** 2) * self.height
        self.surface_area = 2 * math.pi * self.radius * (self.radius + self.height)
        return self


def make_crypto_functions(key, iv):

    def encrypt(plainText):
        cipher = AES.new(key, AES.MODE_CBC, iv)
        return cipher.encrypt(plainText)

    def decrypt(encryptedText):
        cipher = AES.new(key, AES.MODE_CBC, iv)
        return cipher.decrypt(encryptedText)
    return (encrypt, decrypt)


def random_name(**args):
    gender = args['gender']
    region = args['region']
    req = requests.get('https://uinames.com/api/',
                       params={'gender': gender, 'region': region, 'amount': 1})
    if 'error' in req.json():
        raise ValueError(req.text)
    return '{},{}'.format(req.json()['surname'], req.json()['name'])
